### ISSUE TITLE
[Bug]: Plugin does not work after update to 3.4.0.0 from 3.3.2.0

### ISSUE THREAD

**Issue by AnotherStranger** [lyarenei/jellyfin-plugin-listenbrainz#85]
### Bug description

I tried to install the plugin for the first time today. However, the installation did not work.
After downgrading the plugin to Version 3.3.2.0 it did work.

It seems like the MetaBrainz.MusicBrainz.dll cannot be found.

Jellyfin Version: 10.8.13
Broken Version: 3.4.0.0
Working Version: 3.3.2.0

### Steps to reproduce

1. Create a new Jellyfin Server using the official Docker image.
2. Install the plugin using the Repository method
3. Restart jellyfin
4. Observe that the plugin won't start.

### Expected behavior

The plugin should start correctly.

### Actual behavior

The plugin does not start. It goes into a malfunctioning state.
![grafik](https://github.com/lyarenei/jellyfin-plugin-listenbrainz/assets/6563442/1e119329-728d-416b-8291-d45b3b0a7a23)

### Jellyfin logs

```shell
[2024-02-19 07:26:38.989 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "MetaBrainz.Common, Version=3.0.0.0, Culture=neutral, PublicKeyToken=be12cbb629c03021" from "/config/plugins/ListenBrainz_3.4.0.0/MetaBrainz.Common.dll"
[2024-02-19 07:26:39.202 +00:00] [ERR] [1] Emby.Server.Implementations.Plugins.PluginManager: Failed to load assembly "/config/plugins/ListenBrainz_3.4.0.0/MetaBrainz.MusicBrainz.dll". This error occurs when a plugin references an incompatible version of one of the shared libraries. Disabling plugin.
System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
```

### Plugin version

3.4.0

### Jellyfin Version

10.8.13

### Additional info

Thank you for your Help!
Please let me know if I can assist you in any way.

**Comment by lyarenei** [comment]
Hmm, this is weird. I just tried clean installation of the plugin on my server and the plugin works without any issues.

While I understand the message and what to usually do, I honestly don't know what the problem could be in your case. The plugin is compiled with .NET6.0, the same as JF 10.8.13, so the dependency DLLs should be compatible with it.

I also have JF deployed in the official docker image (although through truecharts) => we should have nearly the same environment, so that should not be it either.

Here are some things (I can think of) that you can try out. And of course if you did not try them already. Just to be sure, always do the clean installation (below).
- Install the latest version again (to make sure it was not some weird one-off thing)
- Update to 3.4.0.0 after having the 3.3.2.0 version working
- Install the plugin manually (if you don't know how, here's a [guide](https://github.com/lyarenei/jellyfin-plugin-listenbrainz?tab=readme-ov-file#manual-build-and-installation), just skip the compilation section and download the plugin zip from [latest release](https://github.com/lyarenei/jellyfin-plugin-listenbrainz/releases/tag/3.4.0.0))

---

How to do clean installation of the plugin:
1. Uninstall the plugin from admin UI
2. Go to plugin directory ($CONFIG_DIR/plugins)
3. Remove all `ListenBrainz_*` folders, if any
4. Restart Jellyfin
5. Install plugin again (and restart Jellyfin)

**Comment by amarevite** [comment]
Using `lscr.io/linuxserver/jellyfin:latest`, I have the same problem and get the same error. When I realized the plugin wasn't sending listens, I tried checking the plugin settings but it was malfunctioned.

1. I stopped Jellyfin with `docker stop jellyfin`
2. Removed all `plugins/ListenBrains_*` folders
3. Started Jellyfin with `docker start jellyfin`
4. Installed the plugin from the catalog tab
5. Restarted Jellyfin with `docker restart jellyfin`

But the plugin is still malfunctioned.

Then I installed `3.3.2.0` from the catalog, uninstalled `3.4.0.0`, and ran `docker restart jellyfin`, which led to this on the main plugins page:

![image](https://github.com/lyarenei/jellyfin-plugin-listenbrainz/assets/35640607/bd9cde26-629c-40da-859c-20cdd505c6db)

The plugin works while in this state (still running `3.3.2.0` if I understand correctly), but upon restarting I am back to the single `3.4.0.0` plugin which is still malfunctioned.

**Comment by lyarenei** [comment]
Okay, so this will probably a bigger issue. I have no idea what's going on and worse, I can't reproduce the issue. Or at least right now.

For now, I'll just remove the release from repository and I'll try to look into it later (it's getting too late here and I gotta get some sleep :slightly_smiling_face:).

---

So whoever is reading this issue and has the same problem, please uninstall the `3.4.0.0` version and reinstall version `3.3.2.0`, which seems to work (it should now show as latest version in the repository). Thank you and sorry for the inconvenience.

**Comment by AnotherStranger** [comment]
Hey  @lyarenei, I just got home from work. I will see what I can provide!

**Comment by lyarenei** [comment]
@AnotherStranger Hi, thanks for the log:

```
System.IO.DirectoryNotFoundException: Could not find a part of the path '/_3.4.0.0/cache.json'
```

For manual upgrade, you need to name the plugin folder `ListenBrainz_3.4.0.0`. Sorry, I forgot to mention that.

Alternatively, if you prefer the repo way, you can switch to my development repo, where I deploy development versions. The repository address is `https://repo.xkrivo.net/jellyfin-dev/manifest.json`.

**Comment by AnotherStranger** [comment]
Hey I installed the plugin using the dev repo. However the DirectoryFoundException persists.
Here I included the output of ls:

```
/config/plugins$ ls
 configurations               Fanart_10.0.0.0        ListenBrainz_3.4.0.0      'Playback Reporting_13.0.0.0'  'Subtitle Extract_2.0.0.0'  'TMDb Box Sets_10.0.0.0'
'Cover Art Archive_7.0.0.0'   ListenBrainz_3.3.2.0  'Open Subtitles_19.0.0.0'   Reports_15.0.0.0               TheTVDB_10.0.0.0            TVmaze_10.0.0.0
```

**Comment by lyarenei** [comment]
@AnotherStranger Huh.. So if you upgrade from previously functioning plugin version, you end up with directory not found and if you install 3.4 right away, you get the DLL error, correct? Is that consisten behavior? maybe something broke with the manual upgrade, I don't know.

Can you please try again? Delete all plugin versions, restart Jellyfin, and install version `3.4.0.1` I recently pushed (in dev repo). I tried strictly locking down the .NET version to 6.0 and upgraded the build action for that one. I really don't think it'll help, but let's see.

Thank you.

**Comment by AnotherStranger** [comment]
Hey, after trying to reproduce the issue on another machine I couldn't. Even though I used the same docker-compose file. I try to look for differences

**Comment by AnotherStranger** [comment]
@lyarenei Okay I think I found something. I deactivated every other plugin and now it works. Now I just have to find out what plugin breaks yours!

**Comment by AnotherStranger** [comment]
@lyarenei I found the plugin breaking ListenBrainz! Its "Cover Art Archive":
![grafik](https://github.com/lyarenei/jellyfin-plugin-listenbrainz/assets/6563442/33243434-6d51-479a-b305-5e52e6e3dc5e)

If i disable this plugin everything works.
Can you reproduce the issue if you install this plugin to your test environment?

**Comment by amarevite** [comment]
I can confirm in my case as well that disabling Covert Art Archive resolves the problem for versions `3.4.0.0` and `3.4.0.1` from the dev repo.

**Comment by lyarenei** [comment]
@AnotherStranger Good find, indeed, after installing CoverArt, the plugin breaks

```
[21:06:01] [INF] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: FFmpeg: /usr/lib/jellyfin-ffmpeg/ffmpeg
[21:06:01] [INF] [1] Emby.Server.Implementations.ApplicationHost: ServerId: 572f60ba183543d68a94abce2e79b090
[21:06:01] [ERR] [1] Emby.Server.Implementations.ApplicationHost: Error creating Jellyfin.Plugin.ListenBrainz.EntryPoint
System.TypeInitializationException: The type initializer for 'MetaBrainz.MusicBrainz.Query' threw an exception.
 ---> System.IO.FileLoadException: Could not load file or assembly 'MetaBrainz.Common.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=be12cbb629c03021'. Could not find or load a specific file. (0x80131621)
File name: 'MetaBrainz.Common.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=be12cbb629c03021'
 ---> System.IO.FileLoadException: Could not load file or assembly 'MetaBrainz.Common.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=be12cbb629c03021'.
```
ListenBrainz is built with `6.0`, however CoverArt with `5.1`. And because CoverArt loaded first, the older version was used by the server and ListenBrainz fails:

```
[2024-02-20 21:05:57.994 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.CoverArtArchive, Version=7.0.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/Cover Art Archive_7.0.0.0/Jellyfin.Plugin.CoverArtArchive.dll"
[2024-02-20 21:05:57.995 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.ListenBrainz.Api, Version=3.4.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/ListenBrainz_3.4.0.0/Jellyfin.Plugin.ListenBrainz.Api.dll"
[2024-02-20 21:05:57.998 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.ListenBrainz, Version=3.4.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/ListenBrainz_3.4.0.0/Jellyfin.Plugin.ListenBrainz.dll"
[2024-02-20 21:05:57.998 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.ListenBrainz.Http, Version=3.4.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/ListenBrainz_3.4.0.0/Jellyfin.Plugin.ListenBrainz.Http.dll"
[2024-02-20 21:05:57.998 +00:00] [ERR] [1] Emby.Server.Implementations.Plugins.PluginManager: Failed to load assembly "/config/plugins/ListenBrainz_3.4.0.0/MetaBrainz.Common.Json.dll". Disabling plugin.
System.IO.FileLoadException: Assembly with same name is already loaded
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at Emby.Server.Implementations.Plugins.PluginManager.LoadAssemblies()+MoveNext()
[2024-02-20 21:05:58.072 +00:00] [ERR] [1] Emby.Server.Implementations.Plugins.PluginManager: Failed to load assembly "/config/plugins/ListenBrainz_3.4.0.0/MetaBrainz.MusicBrainz.dll". This error occurs when a plugin references an incompatible version of one of the shared libraries. Disabling plugin.
System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
```

Sadly, this means that plugins do not have isolated dependencies and instead all dependencies are shared globally. This is definitely a problem. However, I am not sure if this is a dotnet feature or a Jellyfin issue. I'll try to raise an issue there.

Other than that, the fix is clear, revert #78, but I'll need to implement rate limit handling there first.

Thank you for the report and info @AnotherStranger @amarevite

**Comment by AnotherStranger** [comment]
> @AnotherStranger Good find, indeed, after installing CoverArt, the plugin breaks
> 
> ```
> [21:06:01] [INF] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: FFmpeg: /usr/lib/jellyfin-ffmpeg/ffmpeg
> [21:06:01] [INF] [1] Emby.Server.Implementations.ApplicationHost: ServerId: 572f60ba183543d68a94abce2e79b090
> [21:06:01] [ERR] [1] Emby.Server.Implementations.ApplicationHost: Error creating Jellyfin.Plugin.ListenBrainz.EntryPoint
> System.TypeInitializationException: The type initializer for 'MetaBrainz.MusicBrainz.Query' threw an exception.
>  ---> System.IO.FileLoadException: Could not load file or assembly 'MetaBrainz.Common.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=be12cbb629c03021'. Could not find or load a specific file. (0x80131621)
> File name: 'MetaBrainz.Common.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=be12cbb629c03021'
>  ---> System.IO.FileLoadException: Could not load file or assembly 'MetaBrainz.Common.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=be12cbb629c03021'.
> ```
> 
> ListenBrainz is built with `6.0`, however CoverArt with `5.1`. And because CoverArt loaded first, the older version was used by the server and ListenBrainz fails:
> 
> ```
> [2024-02-20 21:05:57.994 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.CoverArtArchive, Version=7.0.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/Cover Art Archive_7.0.0.0/Jellyfin.Plugin.CoverArtArchive.dll"
> [2024-02-20 21:05:57.995 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.ListenBrainz.Api, Version=3.4.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/ListenBrainz_3.4.0.0/Jellyfin.Plugin.ListenBrainz.Api.dll"
> [2024-02-20 21:05:57.998 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.ListenBrainz, Version=3.4.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/ListenBrainz_3.4.0.0/Jellyfin.Plugin.ListenBrainz.dll"
> [2024-02-20 21:05:57.998 +00:00] [INF] [1] Emby.Server.Implementations.Plugins.PluginManager: Loaded assembly "Jellyfin.Plugin.ListenBrainz.Http, Version=3.4.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/ListenBrainz_3.4.0.0/Jellyfin.Plugin.ListenBrainz.Http.dll"
> [2024-02-20 21:05:57.998 +00:00] [ERR] [1] Emby.Server.Implementations.Plugins.PluginManager: Failed to load assembly "/config/plugins/ListenBrainz_3.4.0.0/MetaBrainz.Common.Json.dll". Disabling plugin.
> System.IO.FileLoadException: Assembly with same name is already loaded
>    at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
>    at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
>    at System.Reflection.Assembly.LoadFrom(String assemblyFile)
>    at Emby.Server.Implementations.Plugins.PluginManager.LoadAssemblies()+MoveNext()
> [2024-02-20 21:05:58.072 +00:00] [ERR] [1] Emby.Server.Implementations.Plugins.PluginManager: Failed to load assembly "/config/plugins/ListenBrainz_3.4.0.0/MetaBrainz.MusicBrainz.dll". This error occurs when a plugin references an incompatible version of one of the shared libraries. Disabling plugin.
> System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.
> ```
> 
> Sadly, this means that plugins do not have isolated dependencies and instead all dependencies are shared globally. This is definitely a problem. However, I am not sure if this is a dotnet feature or a Jellyfin issue. I'll try to raise an issue there.
> 
> Other than that, the fix is clear, revert #78, but I'll need to implement rate limit handling there first.
> 
> Thank you for the report and info @AnotherStranger @amarevite 

I have to thank you for your help and this great plugin! Maybe the Cover Art Plugin is willing to extract the clashing library into a separate plugin which you both could use as dependency? Take everything I say with a grain of salt tho. I have no experience on jellyfin development so I don't know whether this is possible.

**Comment by lyarenei** [comment]
> Maybe the Cover Art Plugin is willing to extract the clashing library into a separate plugin which you both could use as dependency?

Not exactly a plugin but a shared library (after all, that code can be used anywhere, not just in Jellyfin plugins). And this is already the case, the JSON library is here: https://github.com/Zastai/MetaBrainz.Common.Json :slightly_smiling_face: 

The problem is not that the code is not shared, but because both CoverArt and ListenBrainz basically say "I want a specific version of this library and I am not willing to compromise" (for various reasons). And so as CoverArt is loaded first by the server, it locks its dependencies and wins the race.

Now, I could recompile the plugin with downgraded version of the library and it would be probably just fine. But that would fix only this specific case. There might be very well another plugin out there which would break in exactly the same way. So it's not really a proper solution, more like a band aid. 

Hope this clears it up for you a bit. :slightly_smiling_face:

**Comment by lyarenei** [comment]
@AnotherStranger @amarevite Just want to let you know that the issue should be fixed with version `3.4.1.0`. Or at least I did not encounter any issues with CoverArt plugin.

Please do not forget to change to stable repository. Let me know if there are any issues, thank you.

**Comment by amarevite** [comment]
I can confirm that ListenBrainz `3.4.1.0` from the stable repo works with Cover Art Archive `7.0.0.0`. I'm seeing no issues on my end. Thanks!
